polkadot/node/network/collator-protocol/src/collator_side/metrics.rs

```diff
@@ -344,32 +439,33 @@ impl CollationTracker {
 	// on the collation state.
 	// Collation is evicted after it expires.
 	pub fn track(&mut self, mut stats: CollationStats) {
-		// Check the state of collation to compute ttl.
-		let ttl = if stats.fetch_latency().is_none() {
-			// Disable the fetch timer, to prevent bogus observe on drop.
+		// Disable the fetch timer, to prevent bogus observe on drop.
+		if stats.fetch_latency().is_none() {
 			if let Some(fetch_latency_metric) = stats.fetch_latency_metric.take() {
 				fetch_latency_metric.stop_and_discard();
 			}
-			// Collation was never fetched, expires ASAP
-			0
-		} else if stats.backed().is_none() {
-			MAX_BACKING_DELAY
-		} else if stats.included().is_none() {
-			// Set expiration date relative to relay parent block.
-			stats.backed().unwrap_or_default() + MAX_AVAILABILITY_DELAY
-		} else {
-			// If block included no reason to track it.
-			return
-		};
+		}
 
-		self.expire
-			.entry(stats.relay_parent_number + ttl)
-			.and_modify(|heads| {
-				heads.insert(stats.head);
-			})
-			.or_insert_with(|| HashSet::from_iter(vec![stats.head].into_iter()));
 		self.entries.insert(stats.head, stats);
 	}
+
+	/// Block numbers for blocks in the tracker that were included before the last finalized block.
+	///
+	/// This is used to avoid sending too many requests to the Chain API.
+	pub fn maybe_finalized_blocks(&self, last_finalized: BlockNumber) -> Vec<BlockNumber> {
```

Comment:
Couldn't you just returned the Vec<(BlockNumber, Hash)> and then you don't have to call `ChainApiMessage::FinalizedBlockHash`, what am I missing here ?

Reply:
I can but why?